### DISCUSSION TITLE
Feature Request: Save Session State

### DISCUSSION THREAD

**Discussion by chaitanyabsprip** [qtile/qtile#2234] [Ideas]
There should be an option to save session state such that when enabled you can quit qtile and when you come back all the windows are open in the same place as when qtile was quit. Also, if possible, the contents of the windows be the same. I dont think this has been implemented in any of the WM or DE.

**Comment by ksladowski** [comment]
I don't believe the WM/DE handles this. Its a restriction of X as far as I know. Wayland might support it?

**Comment by m-col** [comment]
(Moved the issue to discussions to make it easier to find snippet ideas in the future)

This functionality is often buggy and error prone, because it isn't always obvious which window belongs to what, where you would want it (and how to get it there), and what command line to execute to re-launch it. I think it is unlikely Qtile will implement this internally, though in a search online I did find a standalone program that offers this behaviour: https://github.com/johannesjo/linux-window-session-manager

But here is a snippet I put together that may get you partly towards what you want:
```python
import pickle
import psutil
import subprocess
from pathlib import Path

from libqtile import hook, qtile, window

state_file = Path('~/.local/share/qtile/state').expanduser()


@hook.subscribe.shutdown
def _save_session():
    cmds = []
    for c in qtile.windows_map.values():
        if isinstance(c, window.Window):
            try:
                pid = c.window.get_net_wm_pid()
            except Exception:
                continue
            cmds.append(psutil.Process(pid).cmdline())

    with state_file.open('wb') as f:
        pickle.dump(cmds, f)


@hook.subscribe.startup_once
def _restore_session():
    if state_file.exists():
        with state_file.open('rb') as f:
            cmds = pickle.load(f)
        for cmd in cmds:
            subprocess.Popen(cmd)
```

To make windows appear on the right group you could either put together some rules for those groups, or you could save their actual desktops at shutdown (retrieved with `c.window.get_wm_desktop()`) and save those alongside the command lines.

**Reply by Quicksilver151** [reply]
i get an error from that snippet in the that last import line
"Cannot find reference 'window' in '__init__.py | __init__.py'"

**Reply by m-col** [reply]
Did you include the imports at the top?
